core/activitystreak/activitiystreak.go

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) 2023 Gobalsky Labs Limited
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE.VEGA file and at https://www.mariadb.com/bsl11.
+//
+// Change Date: 18 months from the later of the date of the first publicly
+// available Distribution of this version of the repository, and 25 June 2022.
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+package activitystreak
+
+import (
+	"context"
+	"sort"
+
+	"code.vegaprotocol.io/vega/core/events"
+	"code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/libs/num"
+	"code.vegaprotocol.io/vega/logging"
+	"golang.org/x/exp/maps"
+
+	vegapb "code.vegaprotocol.io/vega/protos/vega"
+	eventspb "code.vegaprotocol.io/vega/protos/vega/events/v1"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/mocks.go -package mocks code.vegaprotocol.io/vega/core/activitystreak Broker,MarketsStatsAggregator
+
+type PartyActivity struct {
+	Active                               uint64
+	Inactive                             uint64
+	RewardDistributionActivityMultiplier num.Decimal
+	RewardVestingActivityMultiplier      num.Decimal
+}
+
+func (p PartyActivity) IsActive() bool {
+	return p.Active != 0 && p.Inactive == 0
+}
+
+func (p *PartyActivity) ResetMultipliers() {
+	p.RewardDistributionActivityMultiplier = num.DecimalOne()
+	p.RewardVestingActivityMultiplier = num.DecimalOne()
+}
+
+func (p *PartyActivity) UpdateMultipliers(benefitTiers []*types.ActivityStreakBenefitTier) {
+	if !p.IsActive() {
+		// we are not active, nothing changes
+		return
+	}
+
+	// these are sorted properly already
+	for _, b := range benefitTiers {
+		if p.Active < b.MinimumActivityStreak {
+			break
+		}
+
+		p.RewardDistributionActivityMultiplier = b.RewardMultiplier
+		p.RewardVestingActivityMultiplier = b.VestingMultiplier
+	}
+}
+
+type MarketsStatsAggregator interface {
+	GetMarketStats() map[string]*types.MarketStats
+}
+
+type Broker interface {
```

Comment:
Not sure why this interface is here, there's one in the broker package that we use throughout